service/controller/azure_cluster_controller.go

```diff
@@ -165,10 +170,52 @@ func newAzureClusterResources(config AzureClusterConfig, certsSearcher certs.Int
 		}
 	}
 
+	var clientFactory *client.Factory
+	{
+		c := client.FactoryConfig{
+			CacheDuration:      30 * time.Minute,
+			CredentialProvider: config.CredentialProvider,
+			Logger:             config.Logger,
+		}
+
+		clientFactory, err = client.NewFactory(c)
+		if err != nil {
+			return nil, microerror.Mask(err)
+		}
+	}
+
+	var newDebugger *debugger.Debugger
+	{
+		c := debugger.Config{
+			Logger: config.Logger,
+		}
+
+		newDebugger, err = debugger.New(c)
+		if err != nil {
+			return nil, microerror.Mask(err)
+		}
+	}
+
+	var subnetResource resource.Interface
+	{
+		c := subnet.Config{
+			AzureClientsFactory: clientFactory,
```

Comment:
Not really important here, but I was thinking that we could start injecting only those clients that some resource actually needs.
With the current way we are basically doing service-locator pattern.

Reply:
Just not sure if it is possible, as we need credential name to create the client.

In that case injecting a factory method would probably be cleaner, something like:
```
func (namespace string, name string) DeploymentsClient`
```

Reply:
Yeah, that would be great. The main problem right now is that we need the runtime object (CR) in order to create the clients. We can't create the object graph upfront, you know? It's tricky.

Reply:
> The main problem right now is that we need the runtime object (CR) in order to create the clients. We can't create the object graph upfront, you know? It's tricky.

Yeah, we definitely cannot pass in an already created client. We could use creator functions though, and then we could simply pass in factory functions like `factory.GetDeploymentsClient`.

I remember that this pattern is typical in (good) .NET DI libraries, as an alternative to directly injecting an already created dependency object, you can inject a factory method (in case like we have now, only with DI container doing some reflection magic for you :smile:).

But this is out of scope of this PR :)